tephra-core/src/main/java/org/apache/tephra/TransactionManager.java

```diff
@@ -348,27 +348,30 @@ private void cleanupTimedOutTransactions() {
         }
 
         long currentTime = System.currentTimeMillis();
-        List<Long> timedOut = Lists.newArrayList();
+        Map<Long, InProgressType> timedOut = Maps.newHashMap();
```

Comment:
I am just not sure the correctness of that. Probably we can keep the existing way, and have another PR just focus on lock (not necessarily target for perf improvement, but rather for the correctness).